geth/node/peerpool.go

```diff
@@ -0,0 +1,247 @@
+/*
+Peer pool works in a following way:
+1. For each configured topic it will create discovery v5 search query.
+Search query will periodically do regular kademlia lookups with bucket size 16
+and send a topic query to every node that is returned from kademlia lookup.
+Eventually nodes with required topics will be found and we will pass them to p2p server.
+2. Additional loop will be created for every topic that will synchronize
+found nodes with p2p server. This loop will follow next logic:
+- if node is found and max limit of peers is not reached we will add this node to
+  server and assume that it is connected
+- if max limit is reached we will add peer to our peer topic table for later use
+- if min limit is reached - frequency will be changed to a keepalive timer, this is required cause we need
+  frequent lookups only when we are looking for a peer
+3. when peer is disconnected we do 3 things:
+  - select new peer from peers table that was updated no longer than foundTimeout and not
+    in the connected (90s)
+  - set a peer as not connected
+  - check how many peers do we have and in case if we went below min limit - set period to fastSync
+*/
+
+package node
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/discv5"
+)
+
+var (
+	// ErrDiscv5NotRunning returned when pool is started but discover v5 is not running or not enabled.
+	ErrDiscv5NotRunning = errors.New("Discovery v5 is not running")
+)
+
+const (
+	foundTimeout    = 90 * time.Second
```

Comment:
Maybe using `expiration period` instead of `timeout` would make the intent clearer?

Reply:
thanks, better name for sure